### PR TITLE
Consolidate logic to check for deactivated users.

### PR DESCRIPTION
**Pull request by clokep**
Follow-up to #15624 to consolidate some of the logic checks.

### CODE REVIEW

**Comment by clokep** [comment]
CC @dklimpel -- hopefully this handles your feedback?

File: synapse/rest/client/login.py [hunk 1]
```diff
@@ -280,6 +282,9 @@ async def _do_appservice_login(
             login_submission,
             ratelimit=appservice.is_rate_limited(),
             should_issue_refresh_token=should_issue_refresh_token,
+            # The user represented by an appservice's configured sender_localpart
+            # is not actually created in Synapse.
+            should_check_deactivated=qualified_user_id != appservice.sender,
```
Comment (clokep) [inline synapse/rest/client/login.py:287]: I find this confusing that we don't create a real user row in the database for this user, but this seemed like the least hacky way around this?

File: tests/handlers/test_password_providers.py [hunk 2]
```diff
@@ -162,10 +166,16 @@ class PasswordAuthProviderTests(unittest.HomeserverTestCase):
...
+        self, reactor: MemoryReactor, clock: Clock, homeserver: HomeServer
+    ) -> None:
         # we use a global mock device, so make sure we are starting with a clean slate
         mock_password_provider.reset_mock()
-        super().setUp()
+
+        # The mock password provider doesn't register the users, so ensure they
+        # are registered first.
+        self.register_user("u", "not-the-tested-password")
+        self.register_user("user", "not-the-tested-password")
```
Comment (clokep) [inline tests/handlers/test_password_providers.py:178]: You need to have the user registered *at some point*; the LDAP code lazily registers, but here I just do it manually.

File: tests/handlers/test_password_providers.py [hunk 3]
```diff
@@ -185,33 +195,19 @@ def password_only_auth_provider_login_test_body(self) -> None:
...
 
-        # try a weird username / pass. Honestly it's unclear what we *expect* to happen
-        # in these cases, but at least we can guard against the API changing
-        # unexpectedly
-        channel = self._send_password_login(" USER🙂NAME ", " pASS\U0001F622word ")
-        self.assertEqual(channel.code, HTTPStatus.OK, channel.result)
-        self.assertEqual("@ USER🙂NAME :test", channel.json_body["user_id"])
-        mock_password_provider.check_password.assert_called_once_with(
-            "@ USER🙂NAME :test", " pASS😢word "
-        )
```
Comment (clokep) [inline tests/handlers/test_password_providers.py:202]: I have no idea what this is trying to test -- these aren't valid usernames. I guess we could return a *new* username that *does* work in this case; maybe that would be more realistic?

**Comment by dklimpel** [comment]
> CC @dklimpel -- hopefully this handles your feedback?

Yes, thanks!

File: synapse/handlers/jwt.py [hunk 4]
```diff
@@ -103,16 +102,4 @@ async def validate_login(self, login_submission: JsonDict) -> str:
...
-
-        # If the account has been deactivated, do not proceed with the login
-        # flow.
-        try:
-            deactivated = await self._main_store.get_user_deactivated_status(user_id)
-        except StoreError:
-            # JWT lazily creates users, so they may not exist in the database yet.
-            deactivated = False
-        if deactivated:
-            raise UserDeactivatedError("This account has been deactivated")
```
Comment (DMRobertson) [inline synapse/handlers/jwt.py:116]: Can you spell out why this is safe to remove? Presumably it's replaced by similar logic on another path.

File: tests/handlers/test_password_providers.py (hunk 3 above)
Comment (DMRobertson) [reply tests/handlers/test_password_providers.py:202]: I think nuke this and move on with our lives.

**Review by DMRobertson — APPROVED** [review]

**Pull request merged**